arangod/IResearch/IResearchDataStore.cpp

```diff
@@ -1578,49 +1559,47 @@ bool IResearchDataStore::hasSelectivityEstimate() {
   return false;
 }
 
-IResearchDataStore::Stats IResearchDataStore::statsSynced() const {
+IResearchDataStore::Stats IResearchDataStore::stats() const {
+  if (_metricStats) {
+    return _metricStats->load();
```

Comment:
`_asyncSelf` is not yet locked here. Is it excluded that after check on line 1563 another thread will deallocate metrics and call on line 1564 will trigger null-pointer access?

Reply:
Its fine, _asyncSelf about _dataStore, not about all link members

Reply:
What about the following sequence:
Thread A calls `IResearchLink::unload` (internally calls `shutdownDataStore`)
Thread B calls `stats` on the same Link which is `IResearchDataStore::stats`

Thread B executes line IResearchDataStore.cpp:1563 and `_metricStats` is there so check passes.
Thread A executes line IResearchDataStore.cpp:822 and `reset` completes immediately as there are no locks
Thread A executes `removeStats()` and `_metricStats`  is set to `nullptr`
Thread B executes line 1564 effectively dereferencing null pointer.

Reply:
We call `removeMetrics()` (old `removeStats`) before call `_dataStore.resetDataStore();`
So youare right, we should lock mutex in Guard destroy or call `_asyncSelf->lock()`

Reply:
Fixed